A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,184 @@
+Specifically, gRPC will only support the "*all weights equal*" least request algorithm as
+[defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request).
+
+### Least Request LB Policy
+
+#### xDS API Fields
+
+The xDS `Cluster` resource specifies the load balancing policy to use
+when choosing the endpoint to which each request is sent.  The policy to
+use is specified in the [`lb_policy`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L722).
+Prior to this proposal, gRPC supported two values for this field,
+`ROUND_ROBIN` and `RING_HASH`.  With this proposal, we will add support for an
+additional policy, `LEAST_REQUEST`.
+
+The configuration for the Least Request LB policy is the
+[`least_request_lb_config`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L916).
+The field is optional; if not present, defaults will be assumed for all
+of its values.  gRPC will support the
+[`choice_count`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L346)
+field the same way that Envoy does.
+The
+[`active_request_bias`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L371)
+field will be ignored as it is only used in the "*all weights not equal*" (weighted least request)
+case which will not be supported as part of this proposal.
+
+#### Changes to the `xds_cluster_resolver` Policy
+
+As a result of
+[gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md),
+the `xds_cluster_resolver` policy implementation will need additional logic to understand the setup for `LEAST_REQUEST`
+in addition to `ROUND_ROBIN` and `RING_HASH`.
+For `LEAST_REQUEST`, the policy name will be `"LEAST_REQUEST"`, and the config will be the one for the `least_request_experimental` LB Policy described below.
+When the `xds_cluster_resolver` policy sees `"LEAST_REQUEST"`, it will also assume that the
+locality-picking policy is `weighted_target` and the endpoint-picking policy is `least_request_experimental`.
+
+#### `least_request_experimental` LB Policy
+
+We will implement a `least_request_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
```

Comment:
Even though Envoy has no limitation in place I think it makes sense to have one in gRPC!
If others agree on 10 being a reasonable max choice count we can use that I think.
Capping it using `number of endpoints` may also be a good idea but it possibly could reduce performance in smaller scale scenarios. For example with a `choice_count=3` and only two endpoints `a` and `b` where `b` performs much worse than `a` (i.e. much more outstanding requests), the probability of picking the bad one `b` goes from `25%` to `12.5%` if we still allow three choices instead of two (since sampling is independently random).

Reply:
I thought about this a bit more and decided to go ahead and update with max 10 now already.
Happy to update again if it makes sense to change the max value or also take `number of endpoints` into account.